### PR TITLE
Fix unsound C++ union copy-constructor

### PR DESCRIPTION
**Pull request by emilk**
### What
* Closes https://github.com/rerun-io/rerun/issues/3864

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3892) (if applicable)
* [X] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3892)
- [Docs preview](https://rerun.io/preview/bd53891896e118c00fa8d48bf51f5f12eadae453/docs) 
- [Examples preview](https://rerun.io/preview/bd53891896e118c00fa8d48bf51f5f12eadae453/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_types_builder/src/codegen/cpp/mod.rs [hunk 1]
```diff
@@ -895,21 +901,47 @@ impl QuotedObject {
...
+                quote! {
+                    #NEWLINE_TOKEN
+                    #NEWLINE_TOKEN
+                    #comment
+                    #pascal_case_ident(const #pascal_case_ident& other) : _tag(other._tag) {
+                        switch (other._tag) {
+                            #(#placement_new_arms)*
+
+                            case detail::#tag_typename::NONE:
+                                break;
```
Comment (Wumpf) [inline crates/re_types_builder/src/codegen/cpp/mod.rs:925]: this is subtly different to the other two cases where NONE is still doing the trivial memcpy path. _Probably_ not a problem but makes me nervous!
I'd sleep better if the other two cases also do not do a memcpy either.

**Review by Wumpf — APPROVED** [review]
Looking good, but I feel like this warrants a dedicated unit test using one of the test types?

**Comment by emilk** [comment]
> Looking good, but I feel like this warrants a dedicated unit test using one of the test types?

How would you unit-test against UB without a sanatizer though?

* https://github.com/rerun-io/rerun/issues/3893

File: crates/re_types_builder/src/codegen/cpp/mod.rs (hunk 1 above)
Comment (emilk) [reply crates/re_types_builder/src/codegen/cpp/mod.rs:925]: it's fine do either do a memcpy or not in the NONE case - it doesn't matter. I just changed it here to make the generated code for this common case a little nicer to read

**Comment by Wumpf** [comment]
> How would you unit-test against UB without a sanatizer though?

we can't. but for me this caused deterministic failure, so covering that code by a test would be good

**Pull request merged**